File: backend/src/main/java/com/pgvaale/backend/controller/MaidAuthController.java
package com.pgvaale.backend.controller;

import com.pgvaale.backend.entity.Maid;
import com.pgvaale.backend.repository.MaidRepository;
import com.pgvaale.backend.security.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import jakarta.servlet.http.HttpServletRequest;
import java.util.Map;
import java.util.Optional;

@CrossOrigin(origins = "http://localhost:3000")
@RestController
@RequestMapping("/api/maid")
public class MaidAuthController {
    
    @Autowired
    private AuthenticationManager authenticationManager;
    
    @Autowired
    private MaidRepository maidRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;
    
    @Autowired
    private JwtUtil jwtUtil;

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, Object> requestData, HttpServletRequest request) {
        try {
            // Extract required fields
            String username = (String) requestData.get("username");
            String email = (String) requestData.get("email");
            String password = (String) requestData.get("password");
            String name = (String) requestData.get("name");
            String phoneNumber = (String) requestData.get("phoneNumber");
            String aadhaar = (String) requestData.get("aadhaar");
            String services = (String) requestData.get("services");
            
            // Handle monthlySalary conversion from string to double
            Double monthlySalary = null;
            Object salaryObj = requestData.get("monthlySalary");
            if (salaryObj != null) {
                if (salaryObj instanceof Number) {
                    monthlySalary = ((Number) salaryObj).doubleValue();
                } else if (salaryObj instanceof String) {
                    String salaryStr = (String) salaryObj;
                    if (!salaryStr.trim().isEmpty()) {
                        try {
                            monthlySalary = Double.parseDouble(salaryStr);
                        } catch (NumberFormatException e) {
                            return ResponseEntity.badRequest().body("Invalid salary format");
                        }
                    }
                }
            }
            
            String gender = (String) requestData.get("gender");
            String timing = (String) requestData.get("timing");
            String region = (String) requestData.get("region");

            // Validate required fields
            if (username == null || username.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Username is required");
            }
            if (email == null || email.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Email is required");
            }
            if (password == null || password.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Password is required");
            }
            if (name == null || name.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Name is required");
            }
            if (phoneNumber == null || phoneNumber.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Phone number is required");
            }
            if (aadhaar == null || aadhaar.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Aadhaar is required");
            }

            // Check if username exists
            Optional<Maid> existingMaid = maidRepository.findByUsername(username);
            if (existingMaid.isPresent()) {
                return ResponseEntity.badRequest().body("Username already exists");
            }
            
            // Check if email exists
            Optional<Maid> existingEmail = maidRepository.findByEmail(email);
            if (existingEmail.isPresent()) {
                return ResponseEntity.badRequest().body("Email already exists");
            }
            
            // Create new maid
            Maid maid = Maid.builder()
                    .phoneNumber(phoneNumber)
                    .aadhaar(aadhaar)
                    .services(services)
                    .monthlySalary(monthlySalary != null ? monthlySalary : 0.0)
                    .gender(gender)
                    .timing(timing)
                    .region(region)
                    .build();
            
            // Set inherited fields manually
            maid.setUsername(username);
            maid.setPassword(passwordEncoder.encode(password));
            maid.setEmail(email);
            maid.setName(name);
            
            // Save maid
            Maid savedMaid = maidRepository.save(maid);
            
            return ResponseEntity.ok("Maid registration successful for " + savedMaid.getUsername());
            
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Registration failed: " + e.getMessage());
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> loginData, HttpServletRequest request) {
        try {
            String username = loginData.get("username");
            String password = loginData.get("password");
            
            if (username == null || username.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Username is required");
            }
            if (password == null || password.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Password is required");
            }
            
            // Check if maid exists
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.status(401).body("Invalid credentials");
            }
            
            Maid maid = maidOptional.get();
            
            // Check if maid is approved by admin
            if (!maid.isApproved()) {
                return ResponseEntity.status(401).body("Account is pending admin approval. Please wait for approval.");
            }
            
            Authentication auth = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(username, password)
            );
            
            UserDetails userDetails = (UserDetails) auth.getPrincipal();
            String token = jwtUtil.generateToken(userDetails.getUsername(), "ROLE_MAID");
            
            return ResponseEntity.ok(Map.of("token", token));
        } catch (BadCredentialsException e) {
            return ResponseEntity.status(401).body("Invalid credentials");
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Login failed: " + e.getMessage());
        }
    }
} 

File: PGVaaleDotNetBackend/Controllers/MaidAuthController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using PGVaaleDotNetBackend.DTOs;
using PGVaaleDotNetBackend.Entities;
using PGVaaleDotNetBackend.Services;
using System;
using System.Collections.Generic;

namespace PGVaaleDotNetBackend.Controllers
{
    [ApiController]
    [Route("api/maid")]
    public class MaidAuthController : ControllerBase
    {
        private readonly ILogger<MaidAuthController> _logger;
        private readonly MaidService _maidService;
        private readonly EmailService _emailService;
        private readonly JwtService _jwtService;

        public MaidAuthController(ILogger<MaidAuthController> logger, MaidService maidService, EmailService emailService, JwtService jwtService)
        {
            _logger = logger;
            _maidService = maidService;
            _emailService = emailService;
            _jwtService = jwtService;
        }

        [HttpPost("register")]
        public async Task<IActionResult> Register([FromBody] MaidRegisterRequest request)
        {
            try
            {
                // Validate required fields
                if (string.IsNullOrWhiteSpace(request.Username))
                {
                    return BadRequest("Username is required");
                }
                if (string.IsNullOrWhiteSpace(request.Email))
                {
                    return BadRequest("Email is required");
                }
                if (string.IsNullOrWhiteSpace(request.Password))
                {
                    return BadRequest("Password is required");
                }
                if (string.IsNullOrWhiteSpace(request.Name))
                {
                    return BadRequest("Name is required");
                }
                if (string.IsNullOrWhiteSpace(request.PhoneNumber))
                {
                    return BadRequest("Phone number is required");
                }
                if (string.IsNullOrWhiteSpace(request.Aadhaar))
                {
                    return BadRequest("Aadhaar is required");
                }

                // Check if username exists
                var existingMaid = await _maidService.GetMaidByUsernameAsync(request.Username);
                if (existingMaid != null)
                {
                    return BadRequest("Username already exists");
                }

                // Check if email exists
                var existingEmail = await _maidService.GetMaidByEmailAsync(request.Email);
                if (existingEmail != null)
                {
                    return BadRequest("Email already exists");
                }

                // Create new maid
                var maid = new Maid
                {
                    Username = request.Username,
                    Password = _maidService.HashPassword(request.Password),
                    Email = request.Email,
                    Name = request.Name,
                    PhoneNumber = request.PhoneNumber,
                    Aadhaar = request.Aadhaar,
                    Services = request.Services,
                    MonthlySalary = request.MonthlySalary,
                    Gender = request.Gender,
                    Timing = request.Timing,
                    Region = request.Region,
                    Approved = false, // Default to false, needs admin approval
                    Active = true
                };

                // Save maid
                var savedMaid = await _maidService.SaveMaidAsync(maid);

                // Send pending approval email
                try
                {
                    _emailService.SendWelcomeEmail(savedMaid.Email ?? "", savedMaid.Name ?? "", "Maid");
                }
                catch (Exception emailException)
                {
                    // Log email error but don't fail registration
                    _logger.LogError(emailException, "Failed to send welcome email to {Email}", savedMaid.Email);
                }

                return Ok($"Maid registration successful for {savedMaid.Username}. Your account is pending admin approval.");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Maid registration failed");
                return StatusCode(500, $"Registration failed: {ex.Message}");
            }
        }

        [HttpPost("login")]
        public async Task<IActionResult> Login([FromBody] LoginRequest request)
        {
            try
            {
                _logger.LogInformation("Maid login attempt for username: {Username}", request?.Username);
                
                if (string.IsNullOrWhiteSpace(request?.Username))
                {
                    _logger.LogWarning("Maid login failed: Username is required");
                    return BadRequest("Username is required");
                }
                if (string.IsNullOrWhiteSpace(request?.Password))
                {
                    _logger.LogWarning("Maid login failed: Password is required");
                    return BadRequest("Password is required");
                }

                // Check if maid exists
                var maid = await _maidService.GetMaidByUsernameAsync(request.Username);
                if (maid == null)
                {
                    _logger.LogWarning("Maid login failed: Maid not found for username: {Username}", request.Username);
                    return Unauthorized("Invalid credentials");
                }

                // Check if maid is approved by admin
                if (!maid.Approved)
                {
                    _logger.LogWarning("Maid login failed: Account not approved for username: {Username}", request.Username);
                    return Unauthorized("Account is pending admin approval. Please wait for approval.");
                }

                _logger.LogInformation("Maid found: {Username}, stored password hash: {PasswordHash}", maid.Username, maid.Password);

                // Verify password
                if (!_maidService.VerifyPassword(request.Password, maid.Password ?? ""))
                {
                    _logger.LogWarning("Maid login failed: Password verification failed for username: {Username}", request.Username);
                    return Unauthorized("Invalid credentials");
                }

                _logger.LogInformation("Maid login successful for username: {Username}", request.Username);

                // Generate JWT token using the JWT service
                var token = _jwtService.GenerateToken(maid);

                return Ok(new { token });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Maid login failed");
                return StatusCode(500, $"Login failed: {ex.Message}");
            }
        }


    }
}